File: src/atcoder/problems/edpc_a_memo_rec.cpp
#include <iostream>
#include <vector>
using namespace std;

long long rec(int i, vector<long long> &dp, vector<int> &h) {
  if (dp[i] != 1e10) {
    return dp[i];
  }

  if (i > 1) {
    dp[i] = min(dp[i], rec(i - 2, dp, h) + abs(h[i] - h[i - 2]));
  }

  dp[i] = min(dp[i], rec(i - 1, dp, h) + abs(h[i] - h[i - 1]));

  return dp[i];
}

int main() {
  int n;
  cin >> n;

  vector<int> h(n);
  for (int i = 0; i < n; i++) {
    cin >> h[i];
  }

  vector<long long> dp(n, 1e10);
  dp[0] = 0;
  rec(n - 1, dp, h);
  cout << dp[n - 1] << endl;

  for (int i = 0; i < 100000; i++) {
    cout << 1 << " ";
  }
  cout << endl;

  return 0;
}

File: src/atcoder/problems/edpc_a_memo_rec.py
import sys

# 再帰の上限を増やす
sys.setrecursionlimit(10 ** 9)


def rec(i, h, dp):
    if dp[i] < 10 ** 10:
        return dp[i]

    if i > 1:
        dp[i] = min(dp[i], rec(i - 2, h, dp) + abs(h[i] - h[i - 2]))

    dp[i] = min(dp[i], rec(i - 1, h, dp) + abs(h[i] - h[i - 1]))

    return dp[i]


def main():
    n = int(input())
    h = list(map(int, input().split()))

    dp = [10 ** 10] * n
    dp[0] = 0
    rec(n - 1, h, dp)
    print(dp[n - 1])


main()
